File: packages/php/src/Resources/Webhooks.php
<?php

declare(strict_types=1);

namespace Opensms\Resources;

use Opensms\Models;
use Opensms\Page;
use Opensms\Transport;

/**
 * The `webhooks` resource: endpoints, test events, deliveries and replays, plus signature verification helpers.
 * Accessed as `$client->webhooks`.
 */
final class Webhooks
{
    /** @internal */
    public function __construct(private readonly Transport $transport)
    {
    }

    /**
     * @param array<string, mixed> $params `limit`, `cursor`
     */
    public function list(array $params = []): Page
    {
        return Models::page($this->transport->request('GET', '/v1/webhooks', Models::map($params, Models::PAGE_QUERY, [], true)));
    }

    /**
     * Create an endpoint. `url` must be https. The response includes the
     * signing `secret` (`whsec_...`), shown only once.
     *
     * @param array{url: string, events: list<string>, enabled?: bool} $params
     * @param array{idempotencyKey?: string} $options
     * @return array<string, mixed>
     */
    public function create(array $params, array $options = []): array
    {
        $body = Models::map($params, Models::WEBHOOK, ['url', 'events']);

        return Models::object($this->transport->request('POST', '/v1/webhooks', [], $body, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
        ]));
    }

    /**
     * @return array<string, mixed>
     */
    public function get(string $id): array
    {
        return Models::object($this->transport->request('GET', '/v1/webhooks/' . Models::pathParam($id)));
    }

    /**
     * Full replacement (PUT): `url`, `events` and `enabled` are all required.
     *
     * @param array{url: string, events: list<string>, enabled: bool} $params
     * @param array{idempotencyKey?: string} $options
     * @return array<string, mixed>
     */
    public function update(string $id, array $params, array $options = []): array
    {
        $body = Models::map($params, Models::WEBHOOK, ['url', 'events', 'enabled']);

        return Models::object($this->transport->request('PUT', '/v1/webhooks/' . Models::pathParam($id), [], $body, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
        ]));
    }

    /**
     * @param array{idempotencyKey?: string} $options
     */
    public function delete(string $id, array $options = []): void
    {
        $this->transport->request('DELETE', '/v1/webhooks/' . Models::pathParam($id), [], null, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
        ]);
    }

    /**
     * Queue a `webhook.test` delivery. Returns `{status}`.
     *
     * @param array{idempotencyKey?: string} $options
     * @return array<string, mixed>
     */
    public function test(string $id, array $options = []): array
    {
        return Models::object($this->transport->request('POST', '/v1/webhooks/' . Models::pathParam($id) . '/test', [], null, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
        ]));
    }

    /**
     * @param array<string, mixed> $params `limit`, `cursor`
     */
    public function listDeliveries(string $id, array $params = []): Page
    {
        $query = Models::map($params, Models::PAGE_QUERY, [], true);

        return Models::page($this->transport->request('GET', '/v1/webhooks/' . Models::pathParam($id) . '/deliveries', $query));
    }

    /**
     * Re-queue a delivery. `generation` comes from the delivery; `reason`
     * is 5..1000 characters. Returns `{status}`.
     *
     * @param array{generation: int, reason: string} $params
     * @param array{idempotencyKey?: string} $options
     * @return array<string, mixed>
     */
    public function replayDelivery(string $id, string|int $deliveryId, array $params, array $options = []): array
    {
        $body = Models::map($params, Models::WEBHOOK_REPLAY, ['generation', 'reason']);
        $path = '/v1/webhooks/' . Models::pathParam($id) . '/deliveries/' . Models::pathParam($deliveryId, 'deliveryId') . '/replay';

        return Models::object($this->transport->request('POST', $path, [], $body, [
            'idempotent' => true,
            'idempotencyKey' => Models::idempotencyKey($options),
        ]));
    }

    /**
     * Same as {@see \Opensms\Webhook::verifySignature()}.
     *
     * @param array{toleranceSeconds?: int, now?: int} $options
     */
    public function verifySignature(string $payload, string $header, string $secret, array $options = []): bool
    {
        return \Opensms\Webhook::verifySignature($payload, $header, $secret, $options);
    }

    /**
     * Same as {@see \Opensms\Webhook::constructEvent()}.
     *
     * @param array{toleranceSeconds?: int, now?: int} $options
     * @return array<string, mixed>
     */
    public function constructEvent(string $payload, string $header, string $secret, array $options = []): array
    {
        return \Opensms\Webhook::constructEvent($payload, $header, $secret, $options);
    }
}

File: packages/java/src/main/java/io/opensms/Webhooks.java
package io.opensms;

import io.opensms.models.Page;
import io.opensms.models.StatusResult;
import io.opensms.models.WebhookDelivery;
import io.opensms.models.WebhookEndpoint;
import io.opensms.models.WebhookEvent;

import java.util.Map;

/**
 * Manage webhook endpoints and deliveries, and verify incoming signatures.
 * Accessed as {@code client.webhooks()}. Signature helpers are also available
 * without a client on {@link WebhookSignature}.
 */
public final class Webhooks {

    private final ApiTransport t;

    Webhooks(ApiTransport t) {
        this.t = t;
    }

    /** @return the first page of endpoints. */
    public Page<WebhookEndpoint> list() {
        return list(new ListParams());
    }

    /**
     * @param params paging.
     * @return one page.
     */
    public Page<WebhookEndpoint> list(ListParams params) {
        return t.get("/v1/webhooks", params.toQuery(), ApiTransport.pageOf(WebhookEndpoint.class));
    }

    /**
     * Create an endpoint. The response carries the signing {@code secret} once.
     *
     * @param params url and events.
     * @return the endpoint, including {@code secret}.
     */
    public WebhookEndpoint create(WebhookParams params) {
        return create(params, null);
    }

    /**
     * @param params  url and events.
     * @param options per-call options.
     * @return the endpoint, including {@code secret}.
     */
    public WebhookEndpoint create(WebhookParams params, RequestOptions options) {
        return t.json("POST", "/v1/webhooks", params.toWire(), ApiTransport.idempotencyKey(options),
                ApiTransport.type(WebhookEndpoint.class));
    }

    /**
     * @param id endpoint id.
     * @return the endpoint (no secret).
     */
    public WebhookEndpoint get(String id) {
        return t.get("/v1/webhooks/" + Wire.id(id, "id"), null, ApiTransport.type(WebhookEndpoint.class));
    }

    /**
     * Replace an endpoint ({@code PUT}): {@code url}, {@code events} and {@code enabled} are all required.
     *
     * @param id     endpoint id.
     * @param params the full replacement.
     * @return the updated endpoint.
     */
    public WebhookEndpoint update(String id, WebhookParams params) {
        return update(id, params, null);
    }

    /**
     * @param id      endpoint id.
     * @param params  the full replacement.
     * @param options per-call options.
     * @return the updated endpoint.
     */
    public WebhookEndpoint update(String id, WebhookParams params, RequestOptions options) {
        if (params.enabled() == null) {
            throw new IllegalArgumentException("enabled is required for update (full replacement)");
        }
        return t.json("PUT", "/v1/webhooks/" + Wire.id(id, "id"), params.toWire(),
                ApiTransport.idempotencyKey(options), ApiTransport.type(WebhookEndpoint.class));
    }

    /** @param id endpoint id. */
    public void delete(String id) {
        delete(id, null);
    }

    /**
     * @param id      endpoint id.
     * @param options per-call options.
     */
    public void delete(String id, RequestOptions options) {
        t.json("DELETE", "/v1/webhooks/" + Wire.id(id, "id"), null, ApiTransport.idempotencyKey(options),
                ApiTransport.VOID);
    }

    /**
     * Queue a {@code webhook.test} delivery.
     *
     * @param id endpoint id.
     * @return {@code {status}}.
     */
    public StatusResult test(String id) {
        return test(id, null);
    }

    /**
     * @param id      endpoint id.
     * @param options per-call options.
     * @return {@code {status}}.
     */
    public StatusResult test(String id, RequestOptions options) {
        return t.json("POST", "/v1/webhooks/" + Wire.id(id, "id") + "/test", null,
                ApiTransport.idempotencyKey(options), ApiTransport.type(StatusResult.class));
    }

    /**
     * @param id endpoint id.
     * @return the first page of deliveries.
     */
    public Page<WebhookDelivery> listDeliveries(String id) {
        return listDeliveries(id, new ListParams());
    }

    /**
     * @param id     endpoint id.
     * @param params paging.
     * @return one page.
     */
    public Page<WebhookDelivery> listDeliveries(String id, ListParams params) {
        return t.get("/v1/webhooks/" + Wire.id(id, "id") + "/deliveries", params.toQuery(),
                ApiTransport.pageOf(WebhookDelivery.class));
    }

    /**
     * Replay a delivery.
     *
     * @param id         endpoint id.
     * @param deliveryId delivery id.
     * @param generation the delivery's current {@code generation}.
     * @param reason     5 to 1000 characters, recorded for audit.
     * @return {@code {status}}.
     */
    public StatusResult replayDelivery(String id, long deliveryId, long generation, String reason) {
        return replayDelivery(id, deliveryId, generation, reason, null);
    }

    /**
     * @param id         endpoint id.
     * @param deliveryId delivery id.
     * @param generation the delivery's current {@code generation}.
     * @param reason     5 to 1000 characters.
     * @param options    per-call options.
     * @return {@code {status}}.
     */
    public StatusResult replayDelivery(String id, long deliveryId, long generation, String reason,
                                       RequestOptions options) {
        Map<String, Object> body = Wire.map();
        body.put("generation", generation);
        body.put("reason", Wire.require(reason, "reason"));
        return t.json("POST", "/v1/webhooks/" + Wire.id(id, "id") + "/deliveries/" + deliveryId + "/replay",
                body, ApiTransport.idempotencyKey(options), ApiTransport.type(StatusResult.class));
    }

    /**
     * Verify an {@code X-OpenSMS-Signature} header with the default 300 s tolerance.
     *
     * @param payload raw request body bytes.
     * @param header  the header value.
     * @param secret  endpoint secret ({@code whsec_...}), used verbatim.
     * @return whether the signature is valid and fresh.
     */
    public boolean verifySignature(byte[] payload, String header, String secret) {
        return WebhookSignature.verify(payload, header, secret);
    }

    /**
     * @param payload raw request body as a string (UTF-8).
     * @param header  the header value.
     * @param secret  endpoint secret.
     * @return whether the signature is valid and fresh.
     */
    public boolean verifySignature(String payload, String header, String secret) {
        return WebhookSignature.verify(payload, header, secret);
    }

    /**
     * Verify and parse a delivery.
     *
     * @param payload raw request body bytes.
     * @param header  the header value.
     * @param secret  endpoint secret.
     * @return the parsed event.
     * @throws OpensmsException status 0 with code {@code invalid_signature} or {@code expired_signature}.
     */
    public WebhookEvent constructEvent(byte[] payload, String header, String secret) {
        return WebhookSignature.constructEvent(payload, header, secret);
    }

    /**
     * @param payload raw request body as a string (UTF-8).
     * @param header  the header value.
     * @param secret  endpoint secret.
     * @return the parsed event.
     */
    public WebhookEvent constructEvent(String payload, String header, String secret) {
        return WebhookSignature.constructEvent(payload, header, secret);
    }
}
